### ISSUE TITLE
program destination safegoat

### ISSUE THREAD

**Issue by faxm0dem** [balabit/syslog-ng#374]
The two following conditions can have catastrophic consequences:
- definition of a program destination which is not executable (or other similar problems)
- definition of a program destination which does return immediately

The second item had the following consequence in production: server deadlock after a few hours of runtime. I _think_ there should be some safe-goat mechanism to disable the destination if it doesn't run or exits too fast. What do you think?

**Comment by lbudai** [comment]
I agree :-)

( we were faced with the second problem last week... )

**Comment by faxm0dem** [comment]
What I can think of to test before deactivating the destination:

* fail if `program` is not executable
* fail or warn & disable dest if `program` returns with RC != 0
* maybe set a default throttle (relatively high)

**Comment by faxm0dem** [comment]
Here's a minimal configuration file which makes syslog-ng spin:

```syslog-ng
@version:3.6
@include "scl.conf"

options {
    keep_hostname (yes);
    use_dns(no);
    use_fqdn (no);
    threaded(yes);
};

source s_prog { program ('sleep 1'); };

destination d_debug { pipe("/dev/stdout"); };

log {
  source(s_prog);
  destination(d_debug);
};
```

**Comment by faxm0dem** [comment]
The problem seems to be broader: whenever the child process dies, syslog-ng starts spinning.
IMHO syslog-ng should:

* disable the source for a configurable period
* reinitialize it after that period (rerun the command)
* completely disable the source after a configurable amount of retries

**Comment by hbakken** [comment]
I see a slightly different problem.  When starting syslog-ng with a config like this:
@version:3.6
source s { unix-dgram("/dev/log"); };
destination d { program("/bin/false"); };
log {
    source(s);
    destination(d);
};

I observe that sometimes syslog-ng till send SIGTERM to its own process group, and since syslog-ng does not set its process group, it will go to init, or whoever else is starting it.

This can, of course, have rather drastic consequences.

I've been looking at the code a bit, and I think that the following lines are a bit dodgy:
afprogram_dd_kill_child(AFProgramDestDriver *self)
{
    ...
    killpg(getpgid(self->pid), SIGTERM);
}

Here, I think some safeguarding should be in place to make sure the process group of the program is correct.  If getpgid(pid_t pid) returns -1, I'm not really sure what killpg(-1) will do.  The man page says the behaviour is undefined.

And
afprogram_popen(const gchar *cmdline, GIOCondition cond, pid_t *pid, gint *fd)
{
   ...
   setpgid(0, 0);
   ...
}

setpgid() could conceivably fail.  Not sure what would cause that, though, but I think the return code should be checked, just to make sure.

**Comment by hbakken** [comment]
After a bit more digging, it is (probably) not -1 to killpg() which is the problem, but it seems like getpgid(self->pid) sometimes returns syslog-ng's own process group ID, and in that case we end up killing ourselves, and any other process in that process group.

**Comment by bazsi** [comment]
Hmm there were fixes in this area. Can you check the same code in current
master?

**Comment by hbakken** [comment]
Well, it was refactored into a function, but it didn't address the getpgid(self->pid) == getpgrp() issue.  I would imagine the race is still there.

**Comment by bazsi** [comment]
Hi,

I see that we explicitly handle the getpgid() == -1 case and that there's a
setpgid() after fork() which causes the child process to be in its own
process group.

Isn't the problem in the -1 case? e.g. when the child exits on its own,
getpgid() returns -1, and killpg(-1) is equivalent to kill(1), e.g. killing
the init, which is obviously not good.

So I think the problem should be addressed in current master. I see ways to
improve this code, but before going there it would be nice to nail the root
cause of the problem you are seeing.

Bazsi



-- 
Bazsi

**Comment by hbakken** [comment]
Well, the diff we have here that fixed our problem was to check that getpgid(self->pid) != getpgrp().

I think this case could come up if either a) setpgid(0, 0) fails in the child, or b) the child somehow dies before it calls setpgid(0, 0).

I think a test for this would be good anyway, though, because doing killpg(getpgrp()) is a bad idea anyway...

**Comment by bazsi** [comment]
setpgid() is the first call in that process, so it's probably failing then.
can you confirm that by stracing it?

-- 
Bazsi

**Comment by bazsi** [comment]
but in any case, can you submit the getpgrp() check as a pull request? it
probably make sense to check that anyway.


-- 
Bazsi

**Comment by hbakken** [comment]
Sadly, I cannot reproduce the problem when I strace (it probably messes up timing).  The patch the other team here has used is... a bit ugly, but I'll rewrite it a bit and submit (as soon as I find out how that works, haven't done it before).

**Comment by hbakken** [comment]
Pull request (https://github.com/balabit/syslog-ng/pull/1030) created.

**Comment by Kokan** [comment]
This have been fixed with #1817.
